Sources/Nats/NatsConnection.swift

```diff
@@ -497,16 +495,43 @@ class ConnectionHandler: ChannelInboundHandler {
     }
 
     func close() async throws {
+        self.reconnectTask?.cancel()
+        await self.reconnectTask?.value
         self.state = .closed
         try await disconnect()
         self.fire(.closed)
     }
 
-    func disconnect() async throws {
+    private func disconnect() async throws {
         self.pingTask?.cancel()
         try await self.channel?.close().get()
     }
 
+    func suspend() async throws {
+        self.reconnectTask?.cancel()
+        await self.reconnectTask?.value
+        if self.state == .connected {
+            self.state = .suspended
+            try await disconnect()
+            self.fire(.disconnected)
+        } else {
+            self.state = .suspended
+        }
+    }
+
+    func resume() async throws {
+        guard self.state == .suspended else {
```

Comment:
should we make sure state changes are synchronized to avoid races?

Reply:
Good point, changed 👍

Reply:
do you think checking the state than changing it could also be racy? if we ran them in the event loop we shouldn't need locks or we can lock read-modify if it won't cause deadlocks.

Reply:
So you mean managing the state as tasks in event loop right?

Reply:
yes, using `eventloop.execute { }`

Reply:
I'm really not sure what would be the best approach here, but possibly it would be safer to execute these on event loop.

I took a stab at this, please take a look at the latest changes.